native/dtrevc3.go

```diff
@@ -0,0 +1,862 @@
+
+	// Quick return in case of a workspace query.
+	if lwork == -1 {
+		work[0] = float64(n + 2*n*nb)
+		return m
+	}
+
+	const (
+		nbmin = 8
+		nbmax = 128
+	)
+
+	// Use blocked version of back-transformation if sufficient workspace.
+	// Zero-out the workspace to avoid potential NaN propagation.
+	if howmny == lapack.AllEigVecMulQ && lwork >= n+2*n*nbmin {
+		nb = min((lwork-n)/(2*n), nbmax)
+		impl.Dlaset(blas.All, n, 1+2*nb, 0, 0, work, 1+2*nb)
+	} else {
+		nb = 1
+	}
+
+	// Set the constants to control overflow.
+	ulp := dlamchP
+	smlnum := float64(n) / ulp * dlamchS
+	bignum := (1 - ulp) / smlnum
+
+	// Split work into a vector of column norms and an n×2*nb matrix b.
+	norms := work[:n]
+	ldb := 2 * nb
+	b := work[n : n+n*ldb]
+
+	// Compute 1-norm of each column of strictly upper triangular part of T
+	// to control overflow in triangular solver.
+	norms[0] = 0
+	for j := 1; j < n; j++ {
+		norms[j] = 0
+		for i := 0; i < j; i++ {
+			norms[j] += math.Abs(t[i*ldt+j])
+		}
+	}
+
+	bi := blas64.Implementation()
+
+	var (
+		x [4]float64
+
+		iv int // Index of column in current block.
+		is int
+
+		// ip is used below to specify the real or complex eigenvalue:
+		//  ip == 0, real eigenvalue,
+		//        1, first  of conjugate complex pair (wr,wi),
+		//       -1, second of conjugate complex pair (wr,wi).
+		ip        int
+		iscomplex [nbmax]int // Stores ip for each column in current block.
+	)
+
+	if side == lapack.LeftEigVec {
+		goto leftev
+	}
+
+	// Compute right eigenvectors.
+
+	// For complex right vector, iv-1 is for real part and iv for complex
+	// part. Non-blocked version always uses iv=1, blocked version starts
+	// with iv=nb-1 and goes down to 0 or 1.
+	iv = max(2, nb) - 1
+	ip = 0
+	is = m - 1
+	for ki := n - 1; ki >= 0; ki-- {
+		if ip == -1 {
+			// Previous iteration (ki+1) was second of
+			// conjugate pair, so this ki is first of
+			// conjugate pair.
+			ip = 1
+			continue
+		}
+
+		if ki == 0 || t[ki*ldt+ki-1] == 0 {
+			// Last column or zero on sub-diagonal, so this
+			// ki must be real eigenvalue.
+			ip = 0
+		} else {
+			// Non-zero on sub-diagonal, so this ki is
+			// second of conjugate pair.
+			ip = -1
+		}
+
+		if howmny == lapack.SelectedEigVec {
+			if ip == 0 {
+				if !selected[ki] {
+					continue
+				}
+			} else if !selected[ki-1] {
+				continue
+			}
+		}
+
+		// Compute the ki-th eigenvalue (wr,wi).
+		wr := t[ki*ldt+ki]
+		wi := 0.0
+		if ip != 0 {
+			wi = math.Sqrt(math.Abs(t[ki*ldt+ki-1])) * math.Sqrt(math.Abs(t[(ki-1)*ldt+ki]))
+		}
+		smin := math.Max(ulp*(math.Abs(wr)+math.Abs(wi)), smlnum)
+
+		if ip == 0 {
+			// Real right eigenvector.
+
+			b[ki*ldb+iv] = 1
+			// Form right-hand side.
+			for k := 0; k < ki; k++ {
+				b[k*ldb+iv] = -t[k*ldt+ki]
+			}
+			// Solve upper quasi-triangular system:
+			//  [ T[0:ki,0:ki] - wr ]*X = scale*b.
+			for j := ki - 1; j >= 0; {
+				if j == 0 || t[j*ldt+j-1] == 0 {
```

Comment:
I think it is because you cannot modify the loop variable within the loop body in Fortran?